### PR TITLE
Inefficient wait for all clusteroperators

### PR DESCRIPTION
**Pull request by deads2k**
while looking at https://github.com/openshift/installer/pull/7525 and weighing code complexity versus efficient, this may be worth the inefficiency for understandability.  It's an option, not a requirement and I do recognize that it's closer to the original than the efficient.

### CODE REVIEW

File: cmd/openshift-install/create.go [hunk 1]
```diff
@@ -549,6 +561,54 @@ func waitForInitializedCluster(ctx context.Context, config *rest.Config) error {
...
+	}
+
+	waitErr := wait.PollUntilContextCancel(stabilityContext, 1*time.Second, true, waitForAllClusterOperators(clusterOperatorLister))
+	if waitErr != nil {
+		logrus.Errorf("Error checking cluster operator Progressing status: %q", waitErr)
+		stableOperators, unstableOperators, err := currentOperatorStability(clusterOperatorLister)
+		if err != nil {
+			logrus.Errorf("Error checking final cluster operator Progressing status: %q", err)
+		}
+		logrus.Debug("These cluster operators were stable: [%s]", strings.Join(stableOperators.List(), ", "))
```
Comment (r4f4) [inline cmd/openshift-install/create.go:599]:
```suggestion
		logrus.Debugf("These cluster operators were stable: [%s]", strings.Join(sets.List(stableOperators), ", "))
```

File: cmd/openshift-install/create.go [hunk 2]
```diff
@@ -655,3 +719,62 @@ func checkIfAgentCommand(assetStore asset.Store) {
 		logrus.Warning("An agent configuration was detected but this command is not the agent wait-for command")
 	}
 }
+
+func waitForAllClusterOperators(clusterOperatorLister configlisters.ClusterOperatorLister) func(ctx context.Context) (bool, error) {
+	previouslyStableOperators := sets.String{}
```
Comment (r4f4) [inline cmd/openshift-install/create.go:724]:
```suggestion
	previouslyStableOperators := sets.New[string]()
```

File: cmd/openshift-install/create.go [hunk 3]
```diff
@@ -655,3 +719,62 @@ func checkIfAgentCommand(assetStore asset.Store) {
...
+
+		if len(unstableOperators) == 0 {
+			return true, nil
+		}
+
+		return false, nil
+	}
+}
+
+func currentOperatorStability(clusterOperatorLister configlisters.ClusterOperatorLister) (sets.String, sets.String, error) {
```
Comment (r4f4) [inline cmd/openshift-install/create.go:751]:
```suggestion
func currentOperatorStability(clusterOperatorLister configlisters.ClusterOperatorLister) (sets.Set[string], sets.Set[string], error) {
```

File: cmd/openshift-install/create.go [hunk 4]
```diff
@@ -655,3 +719,62 @@ func checkIfAgentCommand(assetStore asset.Store) {
...
+	clusterOperators, err := clusterOperatorLister.List(labels.Everything())
+	if err != nil {
+		return nil, nil, err // lister should never fail
+	}
+
+	stableOperators := sets.String{}
+	unstableOperators := sets.String{}
```
Comment (r4f4) [inline cmd/openshift-install/create.go:758]:
```suggestion
	stableOperators := sets.New[string]()
	unstableOperators := sets.New[string]()
```

File: cmd/openshift-install/create.go [hunk 5]
```diff
@@ -655,3 +719,62 @@ func checkIfAgentCommand(assetStore asset.Store) {
...
+func waitForAllClusterOperators(clusterOperatorLister configlisters.ClusterOperatorLister) func(ctx context.Context) (bool, error) {
+	previouslyStableOperators := sets.String{}
+
+	return func(ctx context.Context) (bool, error) {
+		stableOperators, unstableOperators, err := currentOperatorStability(clusterOperatorLister)
+		if err != nil {
+			return false, err
+		}
+		if newlyStableOperators := stableOperators.Difference(previouslyStableOperators); len(newlyStableOperators) > 0 {
+			for _, name := range newlyStableOperators.List() {
```
Comment (r4f4) [inline cmd/openshift-install/create.go:732]:
```suggestion
			for _, name := range sets.List(newlyStableOperators) {
```

File: cmd/openshift-install/create.go [hunk 6]
```diff
@@ -655,3 +719,62 @@ func checkIfAgentCommand(assetStore asset.Store) {
...
+				logrus.Debugf("Cluster Operator %s is stable", name)
+			}
+		}
+		if newlyUnstableOperators := previouslyStableOperators.Difference(stableOperators); len(newlyUnstableOperators) > 0 {
+			for _, name := range newlyUnstableOperators.List() {
```
Comment (r4f4) [inline cmd/openshift-install/create.go:737]:
```suggestion
			for _, name := range sets.List(newlyUnstableOperators) {
```

File: cmd/openshift-install/create.go [hunk 7]
```diff
@@ -549,6 +561,54 @@ func waitForInitializedCluster(ctx context.Context, config *rest.Config) error {
...
+
+	waitErr := wait.PollUntilContextCancel(stabilityContext, 1*time.Second, true, waitForAllClusterOperators(clusterOperatorLister))
+	if waitErr != nil {
+		logrus.Errorf("Error checking cluster operator Progressing status: %q", waitErr)
+		stableOperators, unstableOperators, err := currentOperatorStability(clusterOperatorLister)
+		if err != nil {
+			logrus.Errorf("Error checking final cluster operator Progressing status: %q", err)
+		}
+		logrus.Debug("These cluster operators were stable: [%s]", strings.Join(stableOperators.List(), ", "))
+		logrus.Errorf("These cluster operators were not stable: [%s]", strings.Join(unstableOperators.List(), ", "))
```
Comment (r4f4) [inline cmd/openshift-install/create.go:600]:
```suggestion
		logrus.Errorf("These cluster operators were not stable: [%s]", strings.Join(sets.List(unstableOperators), ", "))
```

**Review by r4f4 — COMMENTED** [review]
Suggested some changes to use the new set syntax (using Go generics). Other than that, I think I prefer this version since it's much easier to understand and maintain.
I don't think the "inefficiency" is an issue since all we're doing is to wait for the operators anyway.

**Comment by deads2k** [comment]
the generic set doesn't have an ordered `List` that is very nice.

**Comment by deads2k** [comment]
> the generic set doesn't have an ordered `List` that is very nice.

I see.  Very nice.  Didn't know that was a thing

**Comment by r4f4** [comment]
/cc @patrickdillon 
We might have gotten it this time. It would be good to have you check it as well.

**Comment by r4f4** [comment]
I'm waiting for the CI results to stamp it, but LGTM.

**Comment by r4f4** [comment]
With a deadline of 1 minute for all operators to reach stable in https://github.com/openshift/installer/pull/7538, we can see the output in case of failure
```
time="2023-09-28T20:59:16Z" level=error msg="Error checking cluster operator Progressing status: \"context deadline exceeded\""
time="2023-09-28T20:59:16Z" level=debug msg="These cluster operators were stable: [authentication, baremetal, cloud-controller-manager, cloud-credential, cluster-autoscaler, config-operator, console, control-plane-machine-set, csi-snapshot-controller, dns, image-registry, ingress, insights, kube-apiserver, kube-controller-manager, kube-scheduler, kube-storage-version-migrator, machine-api, machine-approver, machine-config, marketplace, monitoring, network, node-tuning, openshift-apiserver, openshift-controller-manager, operator-lifecycle-manager, operator-lifecycle-manager-catalog, operator-lifecycle-manager-packageserver, service-ca, storage]"
time="2023-09-28T20:59:16Z" level=error msg="These cluster operators were not stable: [etcd, openshift-samples]"
```
/lgtm

**Review by r4f4 — COMMENTED** [review]
/lgtm

File: cmd/openshift-install/create.go [hunk 8]
```diff
@@ -551,6 +561,54 @@ func waitForInitializedCluster(ctx context.Context, config *rest.Config) error {
 	return errors.Wrap(err, "failed to initialize the cluster")
 }
 
+// waitForStableOperators ensures that each cluster operator is "stable", i.e. the
+// operator has not been in a progressing state for at least a certain duration,
+// 30 seconds by default. Returns an error if any operator does meet this threshold
```
Comment (patrickdillon) [inline cmd/openshift-install/create.go:566]:
```suggestion
// 30 seconds by default. Returns an error if any operator does not meet this threshold
```
right?

**Pull request merged**